src/app/vertextool/qgsvertextool.cpp

```diff
@@ -2199,6 +2219,9 @@ void QgsVertexTool::moveVertex( const QgsPointXY &mapPoint, const QgsPointLocato
     if ( QgsWkbTypes::hasZ( dragLayer->wkbType() ) && !pt.is3D() )
       pt.addZValue( defaultZValue() );
 
+    if ( QgsWkbTypes::hasM( dragLayer->wkbType() ) && !pt.isMeasure() )
```

Comment:
Good catch -- that being said, I think we need to move this QgsPoint pt( layerPoint ) out of the if ( addingVertex ) block, and use that both for inserting and moving vertices. If you look at the moveVertex function of say a qgspoint or a qgslinestring, you'll notice that they do rely on the Z and M value of the point. being passed.

Reply:
I don't think we can do that. Otherwise if you snap a 3D point on a 2D point you will overwrite the Z value with default.

Reply:
Fair point :)